core/modules/savers/put.js

```diff
@@ -21,36 +21,62 @@ Select the appropriate saver module and set it up
 var PutSaver = function(wiki) {
 	this.wiki = wiki;
 	var self = this;
+	var uri = this.uri();
 	// Async server probe. Until probe finishes, save will fail fast
 	// See also https://github.com/Jermolene/TiddlyWiki5/issues/2276
-	var req = new XMLHttpRequest();
-	req.open("OPTIONS",encodeURI(document.location.protocol + "//" + document.location.hostname + ":" + document.location.port + document.location.pathname));
-	req.onload = function() {
-		// Check DAV header http://www.webdav.org/specs/rfc2518.html#rfc.section.9.1
-		self.serverAcceptsPuts = (this.status === 200 && !!this.getResponseHeader('dav'));
-	};
-	req.send();
+	$tw.utils.httpRequest({
+		url: uri,
+		type: "OPTIONS",
+		callback: function(err, data, xhr) {
+			// Check DAV header http://www.webdav.org/specs/rfc2518.html#rfc.section.9.1
+			self.serverAcceptsPuts = xhr.status === 200 && !!xhr.getResponseHeader("dav");
+		}
+	});
+	// Retrieve ETag if available
+	$tw.utils.httpRequest({
+		url: uri,
+		type: "HEAD",
+		callback: function(err, data, xhr) {
+			self.etag = xhr.getResponseHeader("ETag");
+		}
+	});
+};
+
+PutSaver.prototype.uri = function() {
+	return encodeURI(document.location.toString().split("#")[0]);
 };
 
-PutSaver.prototype.save = function(text,method,callback) {
+// TODO: in case of edit conflict
+// Prompt: Do you want to save over this? Y/N
+// Merging would be ideal, and may be possible using future generic merge flow
+PutSaver.prototype.save = function(text, method, callback) {
 	if (!this.serverAcceptsPuts) {
 		return false;
 	}
-	var req = new XMLHttpRequest();
-	// TODO: store/check ETags if supported by server, to protect against overwrites
-	// Prompt: Do you want to save over this? Y/N
-	// Merging would be ideal, and may be possible using future generic merge flow
-	req.onload = function() {
-		if (this.status === 200 || this.status === 201) {
-			callback(null); // success
-		}
-		else {
-			callback(this.responseText); // fail
+	var self = this;
+	var headers = { "Content-Type": "text/html;charset=UTF-8" };
+	if (this.etag) {
+		headers["If-Match"] = this.etag;
+	}
+	$tw.utils.httpRequest({
+		url: this.uri(),
+		type: "PUT",
+		headers: headers,
+		data: text,
+		callback: function(err, data, xhr) {
+			if (xhr.status === 200 || xhr.status === 201) {
+				self.etag = xhr.getResponseHeader("ETag");
```

Comment:
@Jermolene @edrex: I believe this is incorrect. In my testing on a separate library, the `PUT` response did not provide an ETag, so I had to perform an additional `HEAD` request for that. Not sure whether that's a general bug (which would mean this never actually worked as intended) or perhaps a server-specific idiosyncrasy.

So I _think_ what we need is something like this:

```javascript
    var conclude = function() {
        self.etag = etag;
        callback(null); // success
    };

    var etag = xhr.getResponseHeader("ETag");
    if (etag) {
        conclude();
        return;
    }

    $tw.utils.httpRequest({
        url: this.uri(),
        type: "HEAD",
        callback: function(err, data, xhr) {
            var etag = xhr.getResponseHeader("ETag");
            if (etag) {
                conclude();
            } else {
                callback("¯\_(ツ)_/¯"); // should never occur?
            }
        }
    });
```

Can either of you confirm this?

Reply:
I can confirm, that there have been problems with the webdav saver and missing etags. I didn't have problems with IIS but apache setups did.

Reply:
discussions have been on the group. I didnt watch them very closely

Reply:
continued in #3073